helix-term/src/ui/editor.rs

```diff
@@ -582,6 +582,57 @@ impl EditorView {
         }
     }
 
+    /// Render bufferline at the top
+    pub fn render_bufferline(editor: &Editor, viewport: Rect, surface: &mut Surface) {
+        let scratch = PathBuf::from(SCRATCH_BUFFER_NAME); // default filename to use for scratch buffer
+        surface.clear_with(
+            viewport,
+            editor
+                .theme
+                .try_get("ui.bufferline.background")
+                .unwrap_or_else(|| editor.theme.get("ui.statusline")),
+        );
+        let mut len = 0usize;
+        for doc in editor.documents() {
+            let fname = doc
+                .path()
+                .unwrap_or(&scratch)
+                .file_name()
+                .unwrap_or_default()
+                .to_str()
+                .unwrap_or_default();
+
+            let style = if view!(editor).doc == doc.id() {
+                editor
+                    .theme
+                    .try_get("ui.bufferline.active")
+                    .unwrap_or_else(|| editor.theme.get("ui.background"))
+            } else {
+                editor
+                    .theme
+                    .try_get("ui.bufferline")
+                    .unwrap_or_else(|| editor.theme.get("ui.statusline"))
+            };
+
+            let text = format!(" {}{} ", fname, if doc.is_modified() { "[+]" } else { "" });
+            let offset = text.len();
+
+            surface.set_stringn(
+                1 + viewport.x + len as u16,
+                viewport.y,
+                text,
+                surface.area.width as usize,
+                style,
+            );
+
+            len += offset;
+
+            if len > surface.area.width as usize {
+                break;
+            }
```

Comment:
The offset calculation for `len` will have to use `text.width()` instead of `text.len()`, but we can simply use the return value of `surface.set_string()` (a tuple `(x, y)` with the end position) which does all of the offset calculation for us:

```suggestion
            let used_width = viewport.x - x; // x is initialized to viewport.x before the loop
            let rem_width = surface.area.width.saturating_sub(used_width);
            x = surface
                .set_stringn(x, viewport.y, text, rem_width as usize, style)
                .0;

            if x >= surface.area.right() {
                break;
            }
```

A better solution would be to have a version of [`Buffer::set_spans()`](https://github.com/helix-editor/helix/blob/a766b32ed197221aca7a9fec3af805b4d04a6f1d/helix-tui/src/buffer.rs#L418) that takes an iterator of `Span`s (a Span is a string + a style). Then the width and offset calculation would be managed for us if we output the buffer names from the iterator. But a plain for loop seems good enough for now 👍🏾.